README.md
# MMM-CommonAlertingProtocol

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Shows public alerts sent via the Common Alerting Protocol

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-CommonAlertingProtocol',
            config: {
                // See below for configurable options
            }
        }
    ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `option1`        | *Required* DESCRIPTION HERE
| `option2`        | *Optional* DESCRIPTION HERE TOO <br><br>**Type:** `int`(milliseconds) <br>Default 60000 milliseconds (1 minute)


CHANGELOG.md
# MMM-CommonAlertingProtocol Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [0.1] - Unreleased

...


node_helper.js
/* 
 * Based on Newsfeed by Michael Teeuw https://michaelteeuw.nl
 * MIT Licensed.
 */

const NodeHelper = require("node_helper");
const Log = require("logger");
const NewsfeedFetcher = require("./newsfeedfetcher");
const xml2js = require("xml2js");
const PiP = require("point-in-polygon");
const lodash = require("lodash");

module.exports = NodeHelper.create({
	// Override start method.
	start: function () {
		Log.log(`Starting node helper for: ${this.name}`);
		this.fetchers = [];
		this.fetchFunction = fetch;
		this.fetchCache = null;
		this.alertDetail = {}; // key: url; value: parsed CAP object
		this.location = null; // will be {lat: Y, lon: X} if configured
	},

	// Override socketNotificationReceived received.
	socketNotificationReceived: function (notification, payload) {
		if (notification === "ADD_FEED") {
			this.createFetcher(payload.feed, payload.config);
		}
	},

	ensureFetchCache: function () {
		if (!this.fetchCache) {
			let builder = require('node-fetch-cache').fetchBuilder;
			let fscache = require('node-fetch-cache').FileSystemCache;
			const options = {
				// cacheDirectory: '/some/path', // defaults to .cache
				ttl: 86400000, // ms
			};
			this.fetchCache = builder.withCache(new fscache(options));
			this.fetchFunction = this.fetchCache;
		}
	},

	/**
	 * Creates a fetcher for a new feed if it doesn't exist yet.
	 * Otherwise it reuses the existing one.
	 * @param {object} feed The feed object
	 * @param {object} config The configuration object
	 */
	createFetcher: function (feed, config) {
		const url = feed.url || "";
		const encoding = feed.encoding || "UTF-8";
		const reloadInterval = feed.reloadInterval || config.reloadInterval || 5 * 60 * 1000;
		let useCorsProxy = feed.useCorsProxy;
		if (useCorsProxy === undefined) useCorsProxy = true;
		if (config.cacheFeed) this.ensureFetchCache();
		if (config.lat && config.lon) {
			this.location = { lat: config.lat, lon: config.lon };
		}

		try {
			new URL(url);
		} catch (error) {
			Log.error("CAP feed Error. Malformed feed url: ", url, error);
			this.sendSocketNotification("FEED_ERROR", { error_type: "MODULE_ERROR_MALFORMED_URL" });
			return;
		}

		let fetcher;
		if (typeof this.fetchers[url] === "undefined") {
			Log.log(`Create new newsfetcher for url: ${url} - Interval: ${reloadInterval}`);
			fetcher = new NewsfeedFetcher(url, reloadInterval, encoding, config.logFeedWarnings, useCorsProxy, this.fetchCache);

			fetcher.onReceive(() => {
				this.processAlerts();
			});

			fetcher.onError((fetcher, error) => {
				Log.error("CAP feed Error. Could not fetch feed: ", url, error);
				let error_type = NodeHelper.checkFetchError(error);
				this.sendSocketNotification("FEED_ERROR", {
					error_type
				});
			});

			this.fetchers[url] = fetcher;
		} else {
			Log.log(`Use existing newsfetcher for url: ${url}`);
			fetcher = this.fetchers[url];
			fetcher.setReloadInterval(reloadInterval);
			fetcher.broadcastItems();
		}

		fetcher.startFetch();
	},

	/**
	 * Ensure we've got the data for all active items in all feeds, then broadcast them
	 */
	processAlerts: function () {
		let asyncs = [];
		// Grab all alert detail
		for (let f in this.fetchers) {
			let items = this.fetchers[f].items();
			for (let i in items) {
				let alert = items[i];
				if (alert.url !== undefined && alert.detail === undefined) {
					Log.log(`Retrieving alert detail for ${alert.url}`);
					alert.detail = [];
					asyncs.push(
						this.fetchFunction(alert.url)
							.then(NodeHelper.checkFetchStatus)
							.then((response) => response.text())
							.then((text) => xml2js.parseStringPromise(text))
							.then((result) => { 
								let detail = result?.alert?.info || [];
								alert.detail.push(...detail);
								return alert;
							})
							.catch((error) => { 
								Log.error("CAP feed Error. Could not fetch detail: ", alert.url, error);
								let error_type = NodeHelper.checkFetchError(error);
								this.sendSocketNotification("FEED_ERROR", {
									error_type
								});
							})
					);
				}
			}
		}
		Promise.all(asyncs)
			.finally(() => { this.filterItems(); })
			.then(() => { this.broadcastFeeds(); });
	},

	/** Applies geo-filtering to the alert items, if configured */
	filterItems: function () {
		if (!this.location) return;
		let latlon = [this.location.lat || 0.0, this.location.lon || 0.0];
		for (let f in this.fetchers) {
			let items = this.fetchers[f].items();
			lodash.remove(items, function (item) {
				// If there are any polygons in the alert detail, then we must be inside at least one of them to pass.
				var seenAny = false;
				for (let detail in item.detail) {
					for (let area in item.detail[detail].area) {
						var poly = item.detail[detail].area[area].polygon;
						if (poly) {
							const parsedPolys = parsePolygons(poly);
							seenAny = true;
							for (let p in parsedPolys) {
								if (PiP(latlon, parsedPolys[p])) {
									return false;
								}
							}
						}
					}
				}
				// But if there are no area detail polygons, let it pass.
				return seenAny;
			});
		}
	},

	/**
	 * Creates an object with all feed items of the different registered feeds,
	 * and broadcasts these using sendSocketNotification.
	 */
	broadcastFeeds: function () {
		const feeds = {};
		for (let f in this.fetchers) {
			feeds[f] = this.fetchers[f].items();
		}
		Log.debug(`CAP: Sending alerts for ${Object.keys(feeds).length} feeds`);
		this.sendSocketNotification("FEED_ITEMS", feeds);
	}
});

/** Converts area polygon strings into a form that point-in-polygon can handle
 * Input: array of strings
 * Output: array of polygons (as returned by parsePolygon; each is an array of arrays)
 */
function parsePolygons(data) {
	if (data instanceof String) {
		return [parsePolygon(data)];
	}
	return lodash.map(data, parsePolygon);
}

/** Converts an area polygon from string into an array of arrays
 * Input example: "-40.866,174.111 -40.863,174.125 -40.911,174.128 -40.938,174.108"
 * Output: [ [-40.866,174.111], [-40.863,174.125], [-40.911,174.128], [-40.938,174.108] ]
 */
function parsePolygon(data) {
	var str = String(data);
	var points = str.split(" ");
	var output = [];
	for (let i in points) {
		var coords = points[i].split(",");
		var thisPoint = [];
		for (let c in coords) {
			let n = Number(coords[c]);
			if (n !== n) {
				// parsing failed; Number conversion returned NaN
				return null;
			}
			thisPoint.push(n);
		}
		output.push(thisPoint);
	}
	return output;
}

newsfeedfetcher.js
/* MagicMirror²
 * Node Helper: Newsfeed - NewsfeedFetcher
 *
 * By Michael Teeuw https://michaelteeuw.nl
 * MIT Licensed.
 */

const stream = require("stream");
const FeedMe = require("feedme");
const iconv = require("iconv-lite");
const { htmlToText } = require("html-to-text");
const Log = require("logger");
const NodeHelper = require("node_helper");
const minipass_pipeline = require('minipass-pipeline');

const ego = "CAP Newsfeed-Fetcher"

/**
 * Responsible for requesting an update on the set interval and broadcasting the data.
 * @param {string} url URL of the news feed.
 * @param {number} reloadInterval Reload interval in milliseconds.
 * @param {string} encoding Encoding of the feed.
 * @param {boolean} logFeedWarnings If true log warnings when there is an error parsing a news article.
 * @param {boolean} useCorsProxy If true cors proxy is used for article url's.
 * @param {function} fetchCache If given, used in place of fetch().
 * @class
 */
const NewsfeedFetcher = function (url, reloadInterval, encoding, logFeedWarnings, useCorsProxy, fetchCache = null) {
	let reloadTimer = null;
	let items = [];
	let reloadIntervalMS = reloadInterval;

	let fetchFailedCallback = function () {};
	let itemsReceivedCallback = function () { };
	let fetchCacheFunction = fetchCache;

	if (reloadIntervalMS < 1000) {
		reloadIntervalMS = 1000;
	}

	/* private methods */

	/**
	 * Request the new items.
	 */
	const fetchNews = () => {
		clearTimeout(reloadTimer);
		reloadTimer = null;
		items = [];

		const parser = new FeedMe();

		parser.on("item", (item) => {
			const title = item.title;
			let description = item.description || item.summary || item.content || "";
			const pubdate = item.pubdate || item.published || item.updated || item["dc:date"];
			const url = item.url || item.link || "";
			const guid = item.guid || "";
			const category = item.category || "";

			if (title && pubdate) {
				const regex = /(<([^>]+)>)/gi;
				description = description.toString().replace(regex, "");
				// Convert HTML entities, codes and tag
				description = htmlToText(description, { wordwrap: false });

				items.push({
					title: title,
					description: description,
					pubdate: pubdate,
					url: url,
					guid: guid,
					category: category,
					useCorsProxy: useCorsProxy
				});
			} else if (logFeedWarnings) {
				Log.warn("Can't parse feed item:");
				Log.warn(item);
				Log.warn(`Title: ${title}`);
				Log.warn(`Description: ${description}`);
				Log.warn(`Pubdate: ${pubdate}`);
			}
		});

		parser.on("end", () => {
			this.broadcastItems();
		});

		parser.on("error", (error) => {
			fetchFailedCallback(this, error);
			scheduleTimer();
		});

		//"end" event is not broadcast if the feed is empty but "finish" is used for both
		parser.on("finish", () => {
			scheduleTimer();
		});

		parser.on("ttl", (minutes) => {
			try {
				// 86400000 = 24 hours is mentioned in the docs as maximum value:
				const ttlms = Math.min(minutes * 60 * 1000, 86400000);
				if (ttlms > reloadIntervalMS) {
					reloadIntervalMS = ttlms;
					Log.info(`${ego}: reloadInterval set to ttl=${reloadIntervalMS} for url ${url}`);
				}
			} catch (error) {
				Log.warn(`${ego}: feed ttl is no valid integer=${minutes} for url ${url}`);
			}
		});

		const nodeVersion = Number(process.version.match(/^v(\d+\.\d+)/)[1]);
		const headers = {
			"User-Agent": `Mozilla/5.0 (Node.js ${nodeVersion}) MagicMirror/${global.version}`,
			"Cache-Control": "max-age=0, no-cache, no-store, must-revalidate",
			Pragma: "no-cache"
		};

		let fetchFunction = fetchCacheFunction || fetch;
		fetchFunction(url, { headers: headers })
			.then(NodeHelper.checkFetchStatus)
			.then((response) => {
				let nodeStream;
				if (response.body instanceof stream.Readable || response.body instanceof minipass_pipeline) {
					nodeStream = response.body;
				} else {
					nodeStream = stream.Readable.fromWeb(response.body);
				}
				nodeStream.pipe(iconv.decodeStream(encoding)).pipe(parser);
			})
			.catch((error) => {
				fetchFailedCallback(this, error);
				scheduleTimer();
			});
	};

	/**
	 * Schedule the timer for the next update.
	 */
	const scheduleTimer = function () {
		clearTimeout(reloadTimer);
		reloadTimer = setTimeout(function () {
			fetchNews();
		}, reloadIntervalMS);
	};

	/* public methods */

	/**
	 * Update the reload interval, but only if we need to increase the speed.
	 * @param {number} interval Interval for the update in milliseconds.
	 */
	this.setReloadInterval = function (interval) {
		if (interval > 1000 && interval < reloadIntervalMS) {
			reloadIntervalMS = interval;
		}
	};

	/**
	 * Initiate fetchNews();
	 */
	this.startFetch = function () {
		fetchNews();
	};

	/**
	 * Broadcast the existing items.
	 */
	this.broadcastItems = function () {
		if (items.length <= 0) {
			Log.info(`${ego}: No items to broadcast yet.`);
			return;
		}
		Log.info(`${ego}: Broadcasting ${items.length} items.`);
		itemsReceivedCallback(this);
	};

	this.onReceive = function (callback) {
		itemsReceivedCallback = callback;
	};

	this.onError = function (callback) {
		fetchFailedCallback = callback;
	};

	this.url = function () {
		return url;
	};

	this.items = function () {
		return items;
	};
};

module.exports = NewsfeedFetcher;


MMM-CommonAlertingProtocol.js
/* global Module */

/* Magic Mirror
 * Module: MMM-CommonAlertingProtocol
 * Based on NewsFeed by Michael Teeuw.
 *
 * By Ross Younger
 * MIT Licensed.
 */


Module.register("MMM-CommonAlertingProtocol", {
	defaults: {
		/* 
		Recommendation R.2. – Polling Frequency
		MetService recommends the CAP Feed is polled at least every five minutes to ensure timely
receipt of all Warnings and Watches, but not polled more frequently than every two minutes.
		*/
		reloadInterval: 5 * 60 * 1000,
		feeds: [
			{
				title: "MetService",
				url: "https://alerts.metservice.com/cap/rss",
				config: {
					// You can override settings in commonConfig on a per-feed basis.
				},
			}
		],
		maxDisplayItems: 0,
		prohibitedWords: [],
		removeStartTags: "",
		removeEndTags: "",
		broadcastAlertUpdates: true,
		showAsList: true,
		animationSpeed: 2500,
		updateInterval: 10000, // How often to update the display when showAsList is false
		hideLoading: false,
		commonConfig: {
			showSourceTitle: true,
			showPublishDate: true,
			showAreaDescription: true,
			showIcon: true,
			showAlertTitle: true,
			showOnset: true,
			showDescription: true,
			lengthDescription: 100,
			truncDescription: true,
			wrapDescription: true,
			/* If you want to display the list horizontally (say, showing the icons only) this can be done with custom CSS:
			  .cap-list li { display: inline-block; }
			*/

		},
		cacheFeed: false, // Intended for development only
		lat: null, // Geo-filter location
		lon: null, // Geo-filter location
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	start: function() {
		Log.info(`Starting module: ${this.name}`);

		this.alertItems = [];
		this.activeItem = 0;

		this.loaded = false;
		this.error = null;

		this.registerFeeds();
	},

	getUrlPrefix: function (item) {
		if (item.useCorsProxy) {
			return `${location.protocol}//${location.host}/cors?url=`;
		} else {
			return "";
		}
	},

	getScripts: function() {
		return ["moment.js"];
	},

	getStyles: function () {
		return ["font-awesome.css", "weather-icons.css", "MMM-CommonAlertingProtocol.css"];
	},

	// Load translations files
	getTranslations: function() {
		return {
			en: "translations/en.json",
		};
	},

	getTemplate: function () {
		return "cap.njk";
	},

	getTemplateData: function () {
		if (this.activeItem >= this.alertItems.length)
			this.activeItem = 0;

		return {
			config: this.config,
			item: this.alertItems[this.activeItem],
			items: this.alertItems,
			loaded: this.loaded,
		};
	},

	registerFeeds: function () {
		for (let feed of this.config.feeds) {
			this.sendSocketNotification("ADD_FEED", {
				feed: feed,
				config: this.config
			});
		}
	},

	socketNotificationReceived: function (notification, payload) {
		if (notification === "FEED_ITEMS") {
			this.generateAlerts(payload);

            if (!this.loaded) {
                if (this.config.hideLoading) {
                    this.show();
                }
				this.updateDom(this.config.animationSpeed);
				this.scheduleUpdateInterval();
            }

            this.loaded = true;
            this.error = null;

			this.updateDom(100);
		} else if (notification === "FEED_ERROR") {
			this.error = this.translate(payload.error_type);
			this.updateDom(this.config.animationSpeed);
			this.scheduleUpdateInterval();
		}
	},

	/** Map MetService event types into WI icons */
	convertEventType: function (event) {
		let s = String(event);
		switch (s) {
			case "wind":
				return "strong-wind";
			case "localStorm":
				return "thunderstorm";
			default:
				return s;
		}
	},

	/**
	 * Generate a merged config block for a feed
	 * @note If multiple feeds are configured with the same URL, the results may be surprising.
	 * The helper assumes feed URLs are unique.
	 */
	configForFeed: function (url) {
		for (let iter in this.config.feeds) {
			const feed = this.config.feeds[iter];
			if (feed.url === url) {
				return { ...this.defaults.commonConfig, ... this.config.commonConfig, ...feed.config };
			}
		}
		console.log(`Missing feed config?! ${url}`);
		return { ...this.defaults.commonConfig, ... this.config.commonConfig };
	},

	/**
	 * Generate an ordered list of items for this configured module.
	 * @param {object} feeds An object with feeds returned by the node helper.
	 */
	generateAlerts: function (feeds) {
		let newsItems = [];
		for (let feed in feeds) {
			const feedItems = feeds[feed];
			const thisFeedConfig = this.configForFeed(feed);
			if (this.subscribedToFeed(feed)) {
				for (let item of feedItems) {
					item.sourceTitle = this.titleForFeed(feed);
					item.config = thisFeedConfig;
					if (!(this.config.ignoreOldItems && Date.now() - new Date(item.pubdate) > this.config.ignoreOlderThan)) {
						newsItems.push(item);
					}
				}
			}
		}
		newsItems.sort(function (a, b) {
			const dateA = new Date(a.pubdate);
			const dateB = new Date(b.pubdate);
			return dateB - dateA;
		});

		if (this.config.maxDisplayItems > 0) {
			newsItems = newsItems.slice(0, this.config.maxDisplayItems);
		}

		if (this.config.prohibitedWords.length > 0) {
			newsItems = newsItems.filter(function (item) {
				for (let word of this.config.prohibitedWords) {
					if (item.title.toLowerCase().indexOf(word.toLowerCase()) > -1) {
						return false;
					}
				}
				return true;
			}, this);
		}
		newsItems.forEach((item) => {
			//Remove selected tags from the beginning of rss feed items (title or description)
			if (this.config.removeStartTags === "title" || this.config.removeStartTags === "both") {
				for (let startTag of this.config.startTags) {
					if (item.title.slice(0, startTag.length) === startTag) {
						item.title = item.title.slice(startTag.length, item.title.length);
					}
				}
			}

			if (this.config.removeStartTags === "description" || this.config.removeStartTags === "both") {
				if (this.isShowingDescription) {
					for (let startTag of this.config.startTags) {
						if (item.description.slice(0, startTag.length) === startTag) {
							item.description = item.description.slice(startTag.length, item.description.length);
						}
					}
				}
			}

			//Remove selected tags from the end of rss feed items (title or description)
			if (this.config.removeEndTags) {
				for (let endTag of this.config.endTags) {
					if (item.title.slice(-endTag.length) === endTag) {
						item.title = item.title.slice(0, -endTag.length);
					}
				}

				if (this.isShowingDescription) {
					for (let endTag of this.config.endTags) {
						if (item.description.slice(-endTag.length) === endTag) {
							item.description = item.description.slice(0, -endTag.length);
						}
					}
				}
			}

			// process data we want to directly report
			item.publishDate = moment(new Date(item.pubdate)).fromNow();
			item.severity = item.detail[0]?.severity; // Minor, Moderate, Severe
			item.iconClass = this.convertEventType(item.detail[0]?.event);
			var areas = [];
			item.detail.forEach((detail) => {
				detail.area.forEach((area) => {
					if (area.areaDesc)
						areas.push(area.areaDesc);
				});
			});
			item.areas = areas.join(", ");
			let onset = item.detail[0]?.onset;
			if (onset) {
				item.onset = moment(new Date(onset)).calendar();
			}
		});

		// get updated news items and broadcast them
		const updatedItems = [];
		newsItems.forEach((value) => {
			if (this.alertItems.findIndex((value1) => value1 === value) === -1) {
				// Add item to updated items list
				updatedItems.push(value);
			}
		});

		// check if updated items exist, if so and if we should broadcast these updates, then lets do so
		if (this.config.broadcastAlertUpdates && updatedItems.length > 0) {
			this.sendNotification("CAP_ALERT_UPDATE", { items: updatedItems });
		}

		this.alertItems = newsItems;
	},

	/**
	 * Returns title for the specific feed url.
	 * @param {string} feedUrl Url of the feed
	 * @returns {string} The title of the feed
	 */
		titleForFeed: function (feedUrl) {
			for (let feed of this.config.feeds) {
				if (feed.url === feedUrl) {
					return feed.title || "";
				}
			}
			return "";
	},

	/**
	 * Check if this module is configured to show this feed.
	 * @param {string} feedUrl Url of the feed to check.
	 * @returns {boolean} True if it is subscribed, false otherwise
	 */
	subscribedToFeed: function (feedUrl) {
		for (let feed of this.config.feeds) {
			if (feed.url === feedUrl) {
				return true;
			}
		}
		return false;
	},

	/**
     * Schedule visual update, when in single-item mode
     */
	scheduleUpdateInterval: function () {
		this.updateDom(this.config.animationSpeed);

		// Clear timer if it already exists
		if (this.timer) clearInterval(this.timer);

		if (!this.config.showAsList) {
			this.timer = setInterval(() => {
				if (this.alertItems.length > 1) {
					this.activeItem++;
					this.updateDom(this.config.animationSpeed);
				}
			}, this.config.updateInterval);
		}
	},
});

